File: binary-tree-bfs/average-of-levels-in-binary-tree/average-of-levels-in-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func averageOfLevels(root *TreeNode) []float64 {
	if root == nil {
		return []float64{}
	}

    res := []float64{}
    queue := []*TreeNode{root}

    for len(queue) > 0 {
        size := len(queue)
        sum := 0

        for i := 0; i < size; i++ {
            node := queue[0]
            queue = queue[1:]

			sum += node.Val

            if node.Left != nil { 
				queue = append(queue, node.Left)
			}
            if node.Right != nil { 
				queue = append(queue, node.Right)
			}
        }

        res = append(res, float64(sum) / float64(size))
    }
    return res;
}

File: binary-tree-bfs/average-of-levels-in-binary-tree/average-of-levels-in-binary-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<double> AverageOfLevels(TreeNode root) {
        var result = new List<double>();
        var queue = new Queue<TreeNode>();
        queue.Enqueue(root);

        while (queue.Count > 0)
        {
            int size = queue.Count;
            double sum = 0;

            for (int i = 0; i < size; i++)
            {
                var treeNode = queue.Dequeue();
                sum += treeNode.val;
                if (treeNode.left != null) { queue.Enqueue(treeNode.left); }
                if (treeNode.right != null) { queue.Enqueue(treeNode.right); }
            }

            result.Add(sum / size);
        }
        return result;
    }
}